File: src/go/problems/letter_combinations_test.go
// https://leetcode.com/problems/letter-combinations-of-a-phone-number

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type LetterCombinations struct{}

func TestLetterCombinations(t *testing.T) {
	gen := core.TestSuite[LetterCombinations]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("23").Result([]string{"ad","ae","af","bd","be","bf","cd","ce","cf"})
	}).Add(func(tc *core.TestCase) {
		tc.Param("").Result([]string{})
	}).Run(t)
}

func (LetterCombinations) Solution(digits string) string[] {
    set := [] string {
		"abc", "def",
		"ghi", "jkl", "mno",
		"pqrs", "tuv", "wxyz",
	};
    temp := make([]byte, len(digits))
	result := []string{}
	
	var search func(digits []byte, digitsIndex int, tempIndex int);
	search = func(digits []byte, digitsIndex int, tempIndex int){
		if digitsIndex >= len(digits) {
			result = append(result, string(temp))
			return
		}

		for i := 0; i < len(set[digits[digitsIndex] - '2']); i++ {
			temp[tempIndex] = set[digits[digitsIndex] - '2'][i]
			search(digits, digitsIndex + 1, tempIndex + 1);
		}
	}

	if len(digits) > 0 {
		search([]byte(digits), 0, 0);
	}

	return result
}

File: src/python/problems/letter_combinations_test.py
#https://leetcode.com/problems/letter-combinations-of-a-phone-number

from collections import defaultdict
from core.problem_base import *

class LetterCombinations(ProblemBase):
    def Solution(self, digits: str) -> list[str]:
        if len(digits) <= 0:
            return []
        
        map = {
            '2':"abc", '3':"def",
            '4':"ghi", '5':"jkl", '6':"mno",
            '7':"pqrs", '8':"tuv", '9':"wxyz"
        }
        result = []
        temp = [0] * len(digits)
        def search(digitsIndex: int, tempIndex: int):
            if digitsIndex >= len(digits):
                result.append(''.join(temp))
                return
            
            for ch in map[digits[digitsIndex]]:
                temp[tempIndex] = ch
                search(digitsIndex + 1, tempIndex + 1)
        
        search(0, 0)

        return result
            

if __name__ == '__main__':
    TestGen(LetterCombinations) \
        .Add(lambda tc: tc.Param("23") .Result(["ad","ae","af","bd","be","bf","cd","ce","cf"])) \
        .Run()
